### PR TITLE
Ups nut

### PR DESCRIPTION
**Pull request by arrmo**
#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [ x] Have you signed the [Contributors agreement](http://docs.librenms.org/General/Contributing/)
- [ x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

Changes to correct ups-nut application shortcomings.

Arrgh - by branching now, I get "stuck" with the icon as well ... is that OK?

Thanks!

### CODE REVIEW

**Comment by laf** [comment]
Changing the DS names means existing installs won't work :(

We don't have anything in place to handle this (you'd need to rrdtool dump and re-import) so changing them is out of the question (I think).

**Comment by arrmo** [comment]
Yep, understood - the problem is that the current / legacy install doesn't work ... :(. I started looking, and data is mis-labed and not plotted right.

Thoughts how to deal with it?

Thanks!

**Comment by Rosiak** [comment]
@arrmo 
So what you are saying is, that the current implementation is already broken?

**Comment by arrmo** [comment]
Correct! Unfortunately some mis-mapping of inputs, so what is in stock code right now isn't quite what needs to be there.

**Comment by Rosiak** [comment]
Personally I don't have a problem with merging this, if the current implementation is already broken.
@librenms/reviewers ?

**Comment by laf** [comment]
Can we be 100% certain it's broken for everyone? If so then yes we are ok to merge.

**Comment by arrmo** [comment]
I'm 99% sure, but I definitely don't want to mess anyone else up! That said, some of the graphs included mixed data (input voltage, battery voltage).

Thanks!

**Comment by paulgear** [comment]
I've got nut here, but I'm not monitoring it via applications; is http://docs.librenms.org/Extensions/Applications/#ups-nut the correct way to set up at the moment?  I can give it a try and see what's there.

**Comment by jquagga** [comment]
Could you explain what is broken?  The existing code seems to be working fine for me on two different machines. 

I think this can be added without changing DS names and just adding battery_nominal right?  If you add that after load, keep input voltage and nominal voltage as DS (and change their description) it's just adding a DS to the graphs without removing existing data right?  

If it does need to be changed, I'm not terrible attached to my existing data on UPS battery voltage though and could start over. :-).

**Comment by jquagga** [comment]
Nevermind!  I see why you're doing that. Battery nominal is being loaded as the line voltage. I don't think there is a way around.

**Comment by arrmo** [comment]
Yep, exactly right! That's the issue - the old code didn't crash, but it also didn't quite pull the right data in to the correct plots. As you just noticed (as did I recently) - it's plotting Battery nominal (as nominal) together with line voltage. Two different things ... :-(.

Is there any way you can check my changes? I want to make sure of course that I fixed it correctly ... :). Or just get this merged, and then check - and let me know if anything isn't right? Either way is good with me.

Thanks!

**Comment by jquagga** [comment]
Yeah, I cherry picked the PR and it's running on my install. Of the two machines, I left one with the old data and on the other I deleted the rrd file for nut. It had only been going a day. 

The fresh install is working great it appears. I finally have load reporting which has never been right!  So that's all set. 

On the other machine 2 graphs won't graph due to incompatibilities in the rrd file. Now that's easy enough to fix (I'll just blow that file away too) but everyone will have to do that to get the updated functional graphing going (unless there is a way to migrate data but I didn't care that much about the data). 

I think this is good; with the caveat about existing rrd files.

**Comment by arrmo** [comment]
Agree with you about the rrd files. I just blew my data away as well ... figured it was incorrect data, so not all that much good to me ... :-).

Thanks for checking it out!

File: includes/polling/applications/ups-nut.inc.php [hunk 1]
```diff
@@ -20,36 +20,53 @@
...
 $ups_nut = snmp_get($device, $oid, '-Oqv');
 
+// If "extend" (used above) fails, try "exec" support.
+// Note, exec always splits outputs on newline, so need to use snmp_walk (not a single SNMP entry!)
+if (!$ups_nut) {
+    // Data is in an array, due to how "exec" works with ups-nut.sh output, so snmp_walk to retrieve it
+    $oid = '.1.3.6.1.4.1.2021.7890.2.101';
+    $ups_nut=snmp_walk($device, $oid, '-Oqv');
+}
+//print_r(array_values(explode("\n", $ups_nut)));
```
Comment (paulgear) [inline includes/polling/applications/ups-nut.inc.php]: Can we please get rid of the commented debug lines, or replace them with debug versions?  There are a few more below as well.

**Comment by paulgear** [comment]
So it seems to me that there is about 50% good data in existing RRDs, and 50% not; could we work out a process to dump/restore these RRDs on demand as part of the poll (just before the data_update)?

**Comment by arrmo** [comment]
This one I admit - past my skill set, sorry! Not sure how to do that ... :(.

**Comment by paulgear** [comment]
@arrmo We're not expecting you to do it all on your own.  Let me throw a few ideas around with the team and seem what falls out of the discussion.

**Comment by arrmo** [comment]
Ahh - gotcha. Sure!

Some (much) of the data is valid - it was just in the wrong place. Is there a way to "shuffle" it around? There was some missing data also (not all SNMP data used), but that can just be new data.

Make sense? Thoughts?

Thanks!

**Comment by laf** [comment]
It doesn't look like rrdtool restore supports rrdcached so this won't be possible for people using remote rrdcached with no shared fs.

I'd be happy to merge this, it won't resolve the issue for people with existing data but we should just advise them to delete the rrd file and let is be recreated.

**Comment by arrmo** [comment]
Hi,

Definitely your call, no issue here, but I don't want to mess anyone up. What you say makes sense. I was trying to fix this, not cause grief ... :).

Thanks!

**Comment by murrant** [comment]
Looks like I need to freshen up my rrd file "schema" patch again.  I haven't had much time for coding recently.

**Comment by paulgear** [comment]
Maybe the best path forward is to get out a notification that we'll be pushing a breaking change for NUT UPSes and provide instructions for people to fix it themselves, then give it a couple of weeks, then push it.

**Comment by laf** [comment]
You can't tell people in advance to fix this, it can only be done after they've updated so if you want a notice to go out then it needs to be included in this pull request.

**Comment by arrmo** [comment]
Ummm ... do you need me to do something? I understand your comment, just not sure if I need to do something about it ... :-).

**Comment by paulgear** [comment]
Correct me if I'm wrong, but I believe that until people with existing data dump & restore their RRDs in the correct format, this will prevent polling of NUT UPS entirely.

@laf We've done pre-warnings of breaking changes before, and I think a separate PR released ahead of time is warranted in this case.

**Comment by laf** [comment]
No worries, look forward to seeing the pr

**Comment by jquagga** [comment]
I had restored my install to stock, but I've gone ahead an cherry picked this patch so I could comment on polling after applying.  

After applying the patch, while it looks like the data is still polled by the poller, the graphs stop having data so it seems they're dead.  After blowing away the ups nut rrd files again, it happily starts polling and graphing things properly.  

I'm going to leave this patched code in place.  So of the 13 ups-nut apps listed on stats.librenms.org, at least 2 are already running the new code :).

**Comment by arrmo** [comment]
Hi,

Sorry, not sure if I need to do something here, so we can merge this in - can you clarify?

Thanks!

**Comment by laf** [comment]
@paulgear

**Comment by laf** [comment]
As no one else has moved this forward, I've just created a notification to users informing this pull request will be merged on the 20th Dec.

Any objections speak now please @librenms/reviewers

**Pull request merged**